Optimize computation of gravitational boundary condition

krenzland opened the pull request:
Old way: Use RK to compute the ODE that defines the displacement
New way: Use ADER-magic to compute the Taylor series of the displacement directly and evaluate it at end of timestep and compute integral of displacement. Does this in-place without storing all coefficients. A lot faster, and of arbitrary order.

Draft because: Needs to be tested for realistic application, flop counter needs to be checked (@sebwolf-de if you're bored ;)).

I've already verified that it works for convergence tests. I get the exact same results but faster.

sebwolf-de commented on src/Equations/elastic/Kernels/GravitationalFreeSurfaceBC.h:
```diff
@@ -127,103 +112,82 @@ class GravitationalFreeSurfaceBc {
+    for (unsigned int i = 0; i < nodal::tensor::nodes2D::Shape[0]; ++i) {
+      prevCoefficients[i] = rotatedFaceDisplacement(i, 0);
+      // This is clearly a zeroth order approximation of the integral!
+      integratedDisplacementNodal(i) = deltaTInt * rotatedFaceDisplacement(i, 0);
```
Did you count this flop?

sebwolf-de commented on src/Equations/elastic/Kernels/GravitationalFreeSurfaceBC.h:
```diff
@@ -127,103 +112,82 @@ class GravitationalFreeSurfaceBc {
+      projectKernel.dQ(i) = derivatives + derivativesOffsets[i];
+    }
+
+#ifdef USE_ELASTIC
+    const double rho = materialData.local.rho;
+    const double g = getGravitationalAcceleration(); // [m/s^2]
+    const double Z = std::sqrt(materialData.local.lambda * rho) ;
+#endif
```
What happens else? Should abort, right?

krenzland replied:
It is a standard free surface boundary condition then - this is just s.t. it compiles when using other equations. I'll remove that at some point, but for now I blame our code structure.

sebwolf-de commented on src/Equations/elastic/Kernels/GravitationalFreeSurfaceBC.h:
```diff
@@ -127,103 +112,82 @@ class GravitationalFreeSurfaceBc {
+#ifdef USE_ELASTIC
+        const double curCoeff = uInside - (1.0/Z) * (rho * g * prevCoefficients[i] + pressureInside);
+        // Basically uInside - C_1 * (c_2 * prevCoeff[i] + pressureInside)
+        // 2 add, 2 mul = 4 flops
```
Here you rely on the compiler to precompute `1/Z` and `rho*g`, otherwise I count 6 flops.

krenzland replied:
Yes, but the compiler figures this out. At least it should.

sebwolf-de commented on src/Equations/elastic/Kernels/GravitationalFreeSurfaceBC.h:
```diff
@@ -127,103 +112,82 @@ class GravitationalFreeSurfaceBc {
-    ode::TimeSpan timeSpan = {startTime, startTime + timeStepWidth};
-    auto odeSolverConfig = ode::ODESolverConfig(timeStepWidth);
-    odeSolverConfig.initialDt = timeStepWidth;
+        // 2 * 3 = 6 flops for updating displacement
```
Correct!

sebwolf-de commented on src/Equations/elastic/Kernels/GravitationalFreeSurfaceBC.h:
```diff
@@ -127,103 +112,82 @@ class GravitationalFreeSurfaceBc {
-    odeSolver.setConfig(odeSolverConfig);
-    odeSolver.solve(f, curValue, timeSpan);
+        // 2 flops for updating integral of displacement
+        integratedDisplacementNodal(i) += factorInt * curCoeff;
```
Correct!

sebwolf-de reviewed (commented):
Hi looks good overall, I mostly focused on the flop counters.

krenzland commented:
I've compared the results for a small version of our Palu scenario; they look absolutely identical (in the eyeball norm). This, together with the working convergence tests means that the new implementation is working!
@sebwolf-de I've adjusted the flop counter (by addressing your comments mostly!). Could you have another look please?

sebwolf-de reviewed (approved):
LGTM
